[20.09] Resolve LDDA <-> Library Dataset circular dependency

mvdbeek opened the pull request:
This eliminates a lot of excessive flushes that were needed to persist new `LibraryDatasetDatasetAssociation` (LDDA) and `LibraryDataset` (LD) objects. These intermittent flushes would make LD objects visible before they are associated with a LDDA, causing the Exception in https://github.com/galaxyproject/galaxy/pull/10556#issuecomment-717175641

```
galaxy.web.framework.decorators ERROR 2020-10-27 17:26:25,817 Uncaught exception in exposed API method:
Traceback (most recent call last):
  File "/Users/mvandenb/src/galaxy/lib/galaxy/web/framework/decorators.py", line 294, in decorator
    rval = func(self, trans, *args, **kwargs)
  File "/Users/mvandenb/src/galaxy/lib/galaxy/webapps/galaxy/api/folder_contents.py", line 111, in index
    dataset = content_item.library_dataset_dataset_association.dataset
AttributeError: 'NoneType' object has no attribute 'dataset'
```

The trick here is to use `post_update=True`, so that ldda and ld rows
are INSERTED, followed by an UPDATE to create the link.
This is explained at https://docs.sqlalchemy.org/en/13/orm/relationship_persistence.html.

This is also the groundwork for speeding up library additions by batching flushes.

mvdbeek committed:
Resolve LDDA <-> Library Dataset circular dependency

This eliminates a lot of excessive flushes that were needed to persist
new `LibraryDatasetDatasetAssociation` (LDDA) and `LibraryDataset` (LD) objects.  These
intermittent flushes would make LD objects visible before
they are associated with a LDDA, causing the
Exception in
https://github.com/galaxyproject/galaxy/pull/10556#issuecomment-717175641

```
galaxy.web.framework.decorators ERROR 2020-10-27 17:26:25,817 Uncaught exception in exposed API method:
Traceback (most recent call last):
  File "/Users/mvandenb/src/galaxy/lib/galaxy/web/framework/decorators.py", line 294, in decorator
    rval = func(self, trans, *args, **kwargs)
  File "/Users/mvandenb/src/galaxy/lib/galaxy/webapps/galaxy/api/folder_contents.py", line 111, in index
    dataset = content_item.library_dataset_dataset_association.dataset
AttributeError: 'NoneType' object has no attribute 'dataset'
```

The trick here is to use `post_update=True`, so that ldda and ld rows
are INSERTED, followed by an UPDATE to create the link.
This is explained at
https://docs.sqlalchemy.org/en/13/orm/relationship_persistence.html.

This is also the groundwork for speeding up library additions by
batching flushes.

nsoranzo commented on test/unit/data/test_galaxy_mapping.py:
```diff
@@ -154,6 +155,36 @@ def assert_display_name_converts_to_unicode(item, name):
+        assert target_folder.item_count == 1
+        assert ldda.id
+        assert ldda.library_dataset.id
+        assert ldda.library_dataset_id
```
```suggestion
```

mvdbeek replied:
I've done this intentionally. One of my earlier attempts was a backref only that wouldn't populate `library_dataset_id `

nsoranzo replied:
My eyes saw 2 duplicate lines, sorry!

mvdbeek replied:
That has to be a first time for 🦅 👀 ! 😆

mvdbeek committed:
Refresh folder content once import is complete

The way we did this previously we'd request new folder contents before
the folder item had been added. We can instead request the folder
contents once at the end of the chained import.
